subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy

```diff
@@ -104,6 +104,52 @@ class ZipIntegrationTest extends AbstractIntegrationSpec {
         theZip.hasDescendants('file1.txt', 'file2.txt')
     }
 
+    def zipTaskUseUTF8Encoding() {
+        given:
+        def filename = '中文'
+        def encoding = 'UTF-8'
+        createTestFilesWithEncoding(filename, encoding)
+        buildFile << """
+            task zip(type: Zip) {
+                from 'dir1'
+                from 'dir2'
+                from 'dir3'
+                destinationDir = buildDir
+                archiveName = 'test.zip'
+                encoding = '$encoding'
+            }
+            """
+        when:
+        run 'zip'
+
+        then:
+        def theZip = new ZipTestFixture(file('build/test.zip'), encoding)
+        theZip.hasDescendants("${filename}1.txt", "${filename}2.txt", "${filename}3.txt")
+    }
+
+    def zipTaskUseISOEncoding() {
+        given:
+        def filename = 'ÈÇ'
+        def encoding = 'ISO-8859-1'
+        createTestFilesWithEncoding(filename, encoding)
+        buildFile << """
+            task zip(type: Zip) {
+                from 'dir1'
+                from 'dir2'
+                from 'dir3'
+                destinationDir = buildDir
+                archiveName = 'test.zip'
+                encoding = '$encoding'
+            }
+            """
+        when:
+        run 'zip'
+
+        then:
+        def theZip = new ZipTestFixture(file('build/test.zip'), encoding)
+        theZip.hasDescendants("${filename}1.txt", "${filename}2.txt", "${filename}3.txt")
+    }
+
```

Comment:
Should use Spock `@Unroll` here to avoid test duplication. Would also be useful to have a test explicitly setting `null` encoding.